crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -4551,12 +4551,24 @@ impl<'db> TypeInferenceBuilder<'db> {
             ctx: _,
         } = list;
 
-        for elt in elts {
-            self.infer_expression(elt);
-        }
+        let element_types: Vec<_> = elts
+            .iter()
+            .map(|elt| {
+                let inferred = self.infer_expression(elt);
+                inferred
+                    .literal_fallback_instance(self.db())
+                    .unwrap_or(inferred)
```

Comment:
Calling `literal_fallback_instance` directly will only promote a top-most literal type, so e.g. `[(1,2)]` will be inferred as `list[tuple[Literal[1], Literal[2]]]` instead of `list[tuple[int, int]]`.  To get deep promotion you'll want

```suggestion
                inferred
                    .apply_type_mapping(self.db(), &TypeMapping::PromoteLiterals)
```

(that should help with some of the ecosystem results)

Reply:
Thank you

Reply:
Yeah, I'm just away from my laptop right now, will fix in a couple hours